Docs: Updated Custom Filters

markmandel opened the pull request:
Updates the custom filter documentation to the latest API changes.

This also includes using mdbook's `include` preprocessor to inject the example code into the documentation.

Note: Because of the use of `include` preprocessor I had to remove the Markdown lists, as they wouldn't format correctly
(See: https://github.com/rust-lang/mdBook/issues/1564).

Closes #373

markmandel commented on docs/src/filters/writing_custom_filters.md:
```diff
@@ -3,27 +3,27 @@
 This document provides an overview of the API and how we can go about writing our own [Filters].
 
-#### API Components
+## API Components
```
Just noting: Since I had to remove the lists, I moved the headers to be in level order. PTAL, let me know what you think.

markmandel commented on docs/src/filters/writing_custom_filters.md:
````diff
@@ -38,109 +38,91 @@ These components come together to form the [filter chain].
+```rust,no_run,noplayground
+struct Greet;
+
+impl Filter for Greet {
````
I couldn't include this, so just copied it inline. In theory it could go out of date, but I couldn't think of a better way.

markmandel replied:
This should no longer apply, as now it matches how it was before.

markmandel commented:
Looks like I broke a stack of the docs tests. I'll swing back to this tomorrow and see if I can fix them / if they should be ignored.

iffyio commented on docs/src/filters/writing_custom_filters.md:
````diff
@@ -38,109 +38,91 @@ These components come together to form the [filter chain].
+}
+```
+
+> As a convention we implement a `pub factory() -> DynFilterFactory` function, to 
+> make the next integration step easier.
````
I don't think this is correct? the `factory()` function or constructor wouldn't be visible at all from user code and what's passed into quilkin is the `DynFilterFactory` so it doesn't matter how the object is generated?
```suggestion
```

markmandel replied:
Just so I'm clear, are you saying the description isn't correct? 

If I look at all our Filter implementations, they all have a `pub factory() -> DynFilterFactory` - so I thought it would be good to call out that we conventionally do that, and why.

You are 100% correct that the end use could create a constructor any way, but I thought it would be good to show our usual convention. Maybe the description should be:

> As a convention we generally implement a `pub factory() -> DynFilterFactory` function as a constructor, but a Factory 
> can be constructed as you see fit.

To explain that while you may see this common pattern, it's more by convention, rather than technical limitation. WDYT?

iffyio replied:
Ah, the incorrect part was mostly the 'to make the next integration step easier' bit since the function itself didn't matter.

I think in this case the fact that the quilkin codebase used a function called factory vs anything else doesn't matter at all so it can add confusion to call it out in the first place (as a user I'd find myself reading into what that means and quickly realise it doesn't matter which is wasted time), the user would declare and call the function within their own crate so the function can be called anything or they could construct the factory by some other means (I don't see it as a convention here since both crates don't need to worry about it in the first place).

markmandel replied:
That's fair. Sounds good, I'll remove the comment 👍🏻 

Worst case, if people start asking, we can add it back in 😄

markmandel commented:
Updated, so now it should pass -- basically put the include sections back the way they were before, and updated the include sections so that they have a `// src/main.rs` (or similar) comment like they did previously.

Also updated the initial examples include to use `use quilkin::filters::prelude::*;` - since that is also what we're generally using now as well.
